### PR TITLE
Create an extensible factory interface with universal registry

### PR DESCRIPTION
**Pull request by pursultani**

### CODE REVIEW

File: unchained-api/src/main/java/unchained/MutableConfiguration.java [hunk 1]
```diff
@@ -64,7 +64,7 @@ default C disable(Option<?> option) {
      * @param <E> the type of the value.
      * @return the configuration itself to use for chaining method calls.
      */
-    default <E> C set(Option<E> option, Object value) {
+    default <E> M set(Option<E> option, Object value) {
```
Comment (mmnaseri) [inline unchained-api/src/main/java/unchained/MutableConfiguration.java]: Shouldn't the method signature be `M set(Option<E> option, E value);`?

File: unchained-commons/src/main/java/unchained/commons/contract/DefaultRegistry.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public Set<String> keys() {
+        return Collections.unmodifiableSet(entries.keySet());
+    }
+
+    @Override
+    public Set<T> values() {
+        final Set<T> result = new HashSet<>();
```
Comment (mmnaseri) [inline unchained-commons/src/main/java/unchained/commons/contract/DefaultRegistry.java]: Why can't we do `return Collections.unmodifiableSet(entries.values());`?

File: unchained-commons/src/main/java/unchained/commons/contract/DefaultRegistry.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        result.addAll(entries.values());
+        return Collections.unmodifiableSet(result);
+    }
+
+    static final Map<Class<?>, Registry<? extends Registrable>> ALL = new ConcurrentHashMap<>();
+
+    @SuppressWarnings("unchecked")
+    static <T extends Registrable> Registry<T> lookup(Class<T> type) {
```
Comment (mmnaseri) [inline unchained-commons/src/main/java/unchained/commons/contract/DefaultRegistry.java]: Is this method thread safe?

Comment (pursultani) [reply unchained-commons/src/main/java/unchained/commons/contract/DefaultRegistry.java]: No this method is not thread-safe. However, it should be an atomic operation.

To do so I recommend using `ReentrantReadWriteLock`. The example of `CachedData` in Javadoc fits the bill.

File: unchained-api/src/main/java/unchained/MutableConfiguration.java (hunk 1 above)
Comment (pursultani) [reply unchained-api/src/main/java/unchained/MutableConfiguration.java]: Yes. Copy and paste mistake :stuck_out_tongue:

File: unchained-commons/src/main/java/unchained/commons/contract/DefaultRegistry.java (hunk 2 above)
Comment (pursultani) [reply unchained-commons/src/main/java/unchained/commons/contract/DefaultRegistry.java]: AFAIK `entries.values()` returns `Collection`.

Comment (mmnaseri) [reply unchained-commons/src/main/java/unchained/commons/contract/DefaultRegistry.java]: (that actually doesn't work, so, how about this: `return Collections.unmodifiableSet(new HashSet<>(entries.values()));`)

**Pull request merged**